### PR TITLE
Fix property mangling and implement ES5 compatibility properly

### LINKED ISSUE #440: The package uses ES6 syntax, which causes compatibility issue with older browsers
### Ziggy version

v1.3.1

### Laravel version

v.8.47

## Current Setup

I'm using Ziggy the "regular" way with `@route` in the main blade file.
Ziggy works fine on modern browsers.

## Issue
On chrome 79, the Nullish coalescing operator "??" is not supported. However, the `@route` macro uses it.

Excerpt:
`{...}this.name=t,this.definition=e,this.bindings=e.bindings??{},this.config=r{...}`

Ideally, the javascript code should not use ES6 syntax for better compatibility.

## Question
Is this an issue with my current setup? Or is this an issue with the package itself?

### Ziggy call and context

```js
route('login')
```

### Ziggy configuration

```js
{
  url: "https://luminaire-authentik-erp.test", 
  port: null, 
  defaults: {…}, 
  routes: {…}
}
```

### Route definition

```php
Route::get('login')->name('login');
```

### LINKED ISSUE #441: Unable to use `route().params` after upgrading to 1.3.2
### Ziggy version

1.3.2

### Laravel version

v8.49.0

### Description

Unable to use `route().params` after upgrading to 1.3.2

image
image

image

### Ziggy call and context

```js
route().params
```

### Ziggy configuration

```js
{"url":"http://silver-surfer.test","port":null,"defaults":{},"routes":{"dusk.login":{"uri":"_dusk/login/{userId}/{guard?}","methods":["GET","HEAD"]},"dusk.logout":{"uri":"_dusk/logout/{guard?}","methods":["GET","HEAD"]},"dusk.user":{"uri":"_dusk/user/{guard?}","methods":["GET","HEAD"]},"nova.api.":{"uri":"nova-api/{resource}/{resourceId}/attach-morphed/{relatedResource}","methods":["POST"]},"nova.pages.login":{"uri":"nova/login","methods":["GET","HEAD"]},"nova.login":{"uri":"nova/login","methods":["POST"]},"nova.logout":{"uri":"nova/logout","methods":["POST"]},"nova.pages.password.email":{"uri":"nova/password/reset","methods":["GET","HEAD"]},"nova.password.email":{"uri":"nova/password/email","methods":["POST"]},"nova.pages.password.reset":{"uri":"nova/password/reset/{token}","methods":["GET","HEAD"]},"nova.password.reset":{"uri":"nova/password/reset","methods":["POST"]},"nova.pages.home":{"uri":"nova","methods":["GET","HEAD"]},"nova.pages.403":{"uri":"nova/403","methods":["GET","HEAD"]},"nova.pages.404":{"uri":"nova/404","methods":["GET","HEAD"]},"nova.pages.dashboard":{"uri":"nova/dashboard","methods":["GET","HEAD","POST","PUT","PATCH","DELETE","OPTIONS"]},"nova.pages.dashboard.custom":{"uri":"nova/dashboards/{name}","methods":["GET","HEAD"]},"nova.pages.index":{"uri":"nova/resources/{resourceName}","methods":["GET","HEAD"]},"nova.pages.create":{"uri":"nova/resources/{resourceName}/new","methods":["GET","HEAD"]},"nova.pages.detail":{"uri":"nova/resources/{resourceName}/{resourceId}","methods":["GET","HEAD"]},"nova.pages.edit":{"uri":"nova/resources/{resourceName}/{resourceId}/edit","methods":["GET","HEAD"]},"nova.pages.replicate":{"uri":"nova/resources/{resourceName}/{resourceId}/replicate","methods":["GET","HEAD"]},"nova.pages.lens":{"uri":"nova/resources/{resourceName}/lens/{lens}","methods":["GET","HEAD"]},"nova.pages.attach":{"uri":"nova/resources/{resourceName}/{resourceId}/attach/{relatedResourceName}","methods":["GET","HEAD"]},"nova.pages.edit-attached":{"uri":"nova/resources/{resourceName}/{resourceId}/edit-attached/{relatedResourceName}/{relatedResourceId}","methods":["GET","HEAD"]},"dashboard":{"uri":"dashboard","methods":["GET","HEAD"]},"register":{"uri":"register","methods":["GET","HEAD"]},"login":{"uri":"login","methods":["GET","HEAD"]},"password.request":{"uri":"forgot-password","methods":["GET","HEAD"]},"password.email":{"uri":"forgot-password","methods":["POST"]},"password.reset":{"uri":"reset-password/{token}","methods":["GET","HEAD"]},"password.update":{"uri":"reset-password","methods":["POST"]},"verification.notice":{"uri":"verify-email","methods":["GET","HEAD"]},"verification.verify":{"uri":"verify-email/{id}/{hash}","methods":["GET","HEAD"]},"verification.send":{"uri":"email/verification-notification","methods":["POST"]},"password.confirm":{"uri":"confirm-password","methods":["GET","HEAD"]},"logout":{"uri":"logout","methods":["POST"]}}}
```

### Route definition

```php
Route::prefix('nova')->group(function () {
    Route::get('/{resource}', 'ResourceIndexController@handle');
});
```

### PR DESCRIPTION
**Pull request by bakerkretzmar**
This PR re-introduces the changes from d781b164b8f455730fe8a8b0cbf91f0f8cb55a73 in a way that I'm pretty certain won't break anything this time. It makes Ziggy's default build ES5-compatible.

**TLDR:** internal-only property getters like `get _location()` are mangled by Terser in a way that breaks them, so I changed our one internal getter to a regular method, `_location()`. It's now mangled correctly, so it's not publicly accessible but everything still works.

Closes #441, and closes #440 properly.

---

I'm not 100% sure what the root issue is here, but I'm pretty close and I'm sick enough of bundlers to stop playing with it now. This is what I've found so far:

JavaScript classes like Ziggy's `Router` are transpiled to a combination of objects with custom prototypes and functions to make them ES5 compatible, which looks kind of like this with microbundle's Babel setup:

```js
// Input

class Router {
    current() {
        return this._privateCurrent();
    }
    _privateCurrent() {
        return 'here';
    }
    location() {
        return this._location;
    }
    get _location() {
        return 'there';
    }
}

// Output (lots omitted)

  // ...
  var _proto = Router.prototype;

  _proto.current = function current() {
    return this._privateCurrent();
  };

  _proto._privateCurrent = function _privateCurrent() {
    return 'here';
  };

  _proto.location = function location() {
    return this._location;
  };

  _createClass(Router, [{
    key: "_location",
    get: function get() {
      return 'there';
    }
  }]);
  // ...
```

Notice that the methods become real properties of the Router object (because microbundle uses Babel's `loose` mode), while the `_location` getter is represented by an object with its key and a backing `get` function. Importantly, the properties and methods of the Router object are **not** represented by strings—`current` and `_privateCurrent` are real JavaScript tokens.

Terser, which `microbundle` runs our code through after Babel transpiles it, mangles properties and methods starting with `_` to shorten their names and make them inaccessible publicly. However this mangling obviously ignores strings, since mangling strings, like `'here'` or `'there'`, would destroy the functionality of the code.

Terser's output looks _kind of_ like this, but more optimized and a lot harder to read:

```js
p.current=function current(){return this.cur()}
p.cur=function cur(){return 'here'}
p.location=function location(){return this.loc}
c(Router, [{key:"_location",get:function get(){return 'there'}}])
```

`_privateCurrent` is mangled to `cur` everywhere it appears, shortening it's actual name and all calls to it, and the `location` method's internal `this._location` call is mangled to `this.loc`, but for the same reason that it doesn't mangle `'here'` and `'there'`, Terser **doesn't mangle the `"_location"` key** for that getter! Because its key is a string, `_location` doesn't get renamed properly, which is why `this.loc` doesn't exist when Ziggy later tries to call it from the `current()` method or `params` getter.

There might be a 'better' fix for this but it almost definitely involves setting up our own custom Rollup build, which I don't feel like doing this weekend 😅 for now, making `_location` a method solves this problem.

### CODE REVIEW

**Comment by jakebathman** [comment]
Geez, what a mess. Good detective work to at least get something that works and then we can just _never touch it again_ 😅

**Review by jakebathman — APPROVED** [review]
